### PR TITLE
Support stop or keep of child actors when parent is restarted, #25556

### PR DESCRIPTION
**Pull request by patriknw**
* stash messages and signals while waiting for children to be stopped
* handle watch of other actors
* exception from unstash
* exception from first setup
* merge RestartSupervisor and BackoffSupervisor
* remove unused PreStart
* docs

Refs #25556

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala [hunk 1]
```diff
@@ -197,6 +200,56 @@ sealed abstract class SupervisorStrategy {
...
+   *   if the limit is exceeded the child actor is stopped
+   * @param withinTimeRange duration of the time window for maxNrOfRetries
+   */
+  def withLimit(maxNrOfRetries: Int, withinTimeRange: java.time.Duration): RestartSupervisorStrategy
+
+  /**
+   * Stop or keep child actors when the parent actor is restarted.
+   * @param enabled if `true` then child actors are stopped, otherwise they are kept
+   */
+  def withStopChildren(enabled: Boolean): RestartSupervisorStrategy
```
Comment (hepin1989) [inline akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala]: how about `StopAll` and `StopNone`?

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 2]
```diff
@@ -116,168 +130,217 @@ private class ResumeSupervisor[T, Thr <: Throwable: ClassTag](ss: Resume) extend
...
+        case OptionVal.None ⇒ newBehavior
+        case OptionVal.Some((stashBuffer, _)) ⇒
+          restartingInProgress = OptionVal.None
+          stashBuffer.unstashAll(ctx.asScala.asInstanceOf[scaladsl.ActorContext[Any]], newBehavior.unsafeCast)
       }
-      scheduleRestart(ctx, t)
+      nextBehavior.narrow
+    } catch handleException(ctx, signalRestart = () ⇒ ())
+    // FIXME signal Restart is not done if unstashAll throws, unstash of each message may return a new behavior and
+    //      it's the failing one that should receive the signal
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: will not do it in this PR but added issue https://github.com/akka/akka/issues/26148

File: akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala]: Seems unlikely we will add more states so boolean is fine.

Adding a note about the default would be good though.

File: akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala [hunk 3]
```diff
@@ -197,6 +200,56 @@ sealed abstract class SupervisorStrategy {
...
+
+  /**
+   * While restarting (waiting for children to stop) incoming messages and signals are
+   * stashed, and delivered later to the newly restarted behavior. This property defines
+   * the capacity in number of messages of the stash buffer. If the capacity is exceed
+   * then additional incoming messages are dropped.
+   */
+  def withStashCapacity(capacity: Int): RestartSupervisorStrategy
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala]: Note here about being disabled by default as well.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 4]
```diff
@@ -116,168 +130,217 @@ private class ResumeSupervisor[T, Thr <: Throwable: ClassTag](ss: Resume) extend
...
+  private def increaseRestartCount(): Unit = {
+    strategy match {
+      case restart: Restart ⇒
+        val timeLeft = deadlineHasTimeLeft
+        val newDeadline = if (deadline.isDefined && timeLeft) deadline else OptionVal.Some(Deadline.now + restart.withinTimeRange)
+        restartCount = if (timeLeft) restartCount + 1 else 1
+        deadline = newDeadline
+      case _: Backoff ⇒
+        restartCount += 1
+    }
```
Comment (johanandren) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: Doesn't always increase so maybe `updateRestartCount()` rather?

**Review by johanandren — COMMENTED** [review]
Some small notes, but will need to do a longer review to really grok these changes, big PR.

File: akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala (hunk 3 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/SupervisorStrategy.scala]: Ah, -1 means use config, mention that here.

**Review by johanandren — COMMENTED** [review]
More thorough review done now, looks good, some minor comments.

File: akka-actor-typed/src/main/resources/reference.conf [hunk 5]
```diff
@@ -17,6 +17,12 @@ akka.actor.typed {
 
   # Receptionist is started eagerly to allow clustered receptionist to gather remote registrations early on.
   library-extensions += "akka.actor.typed.receptionist.Receptionist"
+
+  # While an actor is restarted (waiting for backoff to expire and children to stop)
+  # incoming messages and signals are stashed, and delivered later to the newly restarted
+  # behavior. This property defines the capacity in number of messages of the stash
+  # buffer. If the capacity is exceed then additional incoming messages are dropped.
+  restart-stash-capacity = 1000
```
Comment (chbatey) [inline akka-actor-typed/src/main/resources/reference.conf]: Do we deliver the first 1000 or the last 1000 messages?

**Review by chbatey — APPROVED** [review]
Nice

File: akka-actor-typed/src/main/resources/reference.conf (hunk 5 above)
Comment (patriknw) [reply akka-actor-typed/src/main/resources/reference.conf]: the first: " If the capacity is exceed then additional incoming messages are dropped."

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 6]
```diff
@@ -7,32 +7,38 @@ package internal
...
  * INTERNAL API
  */
 @InternalApi private[akka] object Supervisor {
   def apply[T, Thr <: Throwable: ClassTag](initialBehavior: Behavior[T], strategy: SupervisorStrategy): Behavior[T] = {
     strategy match {
-      case r: Resume ⇒
-        Behaviors.intercept[T, T](new ResumeSupervisor(r))(initialBehavior)
       case r: Restart ⇒
         Behaviors.intercept[T, T](new RestartSupervisor(initialBehavior, r))(initialBehavior)
+      case r: Backoff ⇒
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: right, fixed in afaad8f

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala [hunk 7]
```diff
@@ -116,168 +130,217 @@ private class ResumeSupervisor[T, Thr <: Throwable: ClassTag](ss: Resume) extend
...
-        restart(ctx, t)
-        Behavior.validateAsInitial(Behavior.start(initial, ctx))
+      strategy match {
+        case _: Restart ⇒
+          // if unlimited restarts then don't restart if starting fails as it would likely be an infinite restart loop
+          if (strategy.unlimitedRestarts() || ((restartCount + 1) >= strategy.maxRestarts && deadlineHasTimeLeft)) {
+            // don't log here as it'll be logged as ActorInitializationException
+            throw t
+          } else {
+            increaseRestartCount()
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: good catch, off by one err
fixed in b7fcbdc

**Comment by patriknw** [comment]
@johanandren incorporated feedback, please take a look at the changes

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/FaultToleranceDocTest.java [hunk 8]
```diff
@@ -69,6 +69,7 @@ public void bubblingSample() {
           }).build();
     });
 
+    // #bubbling-example
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-actor-typed-tests/src/test/java/jdocs/akka/typed/FaultToleranceDocTest.java]: it was correct but difficult to follow so I moved it to a separate class in 6b48567

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/BubblingSample.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+  public static void main(String[] args) {
+    final ActorSystem<Message> system = ActorSystem.create(bossBehavior, "boss");
+
+    system.tell(new Fail("boom"));
+    // this will now bubble up all the way to the boss and as that is the user guardian it means
+    // the entire actor system will stop
+  }
+}
+// #bubbling-example
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/java/jdocs/akka/typed/BubblingSample.java]: Good, much easier to follow

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala (hunk 2 above)
Comment (helena) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: I will start #26148 after this is merged.

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala]: @helena I merged this

**Pull request merged**